Json output

ehildenb opened the pull request:
@dwightguth this introduces the Json parser, and adds a `KPrint` test which exercises both the `JSON` and `BINARY` parser/unparser. I still need to figure out how to run the `KAST` parser.

Commit:
kernel/.../JsonParser: add JsonParser deserializer

Commit:
kernel/pom.xml: add javax.json Json library as dependency

Commit:
kernel/.../ToJson: add Json serializer to unparsers

Commit:
kernel/.../{KPrint,OutputModes,PrintOptions}: add json output as option

Commit:
kernel/.../BinaryKASTTest: do not create unused files

Commit:
kernel/.../KPrint: isolate strict serializations from pretty prints

ehildenb commented:
Blocked on merging the #117 .

Commit:
kernel/.../JsonParser: helpers `parse` for reading `String` or `byte[]`

Commit:
kernel/.../KPrintTest: test the unparser/parser duos

Commit:
kernel/.../JsonParser: use equals method for string comparison, "token" key for KTOKEN

Commit:
kernel/.../KPrintTest: compare through KAST format instead of directly

Commit:
kernel/.../KPrintTest: enable testing the KAST serializer/deserializer as well

Commit:
kernel/.../KPrintTest: more refined testing of subterms

ehildenb commented:
@dwightguth ready for review.

ehildenb commented:
Jenkins: test this please

dwightguth commented on kernel/src/main/java/org/kframework/parser/json/JsonParser.java:
```diff
@@ -0,0 +1,103 @@
+                             , INJECTEDKLABEL = "InjectedKLabel"
+                             ;
+
+    public static K parse(byte[] data) throws UnsupportedEncodingException {
```
This should really catch UnsupportedEncodingException and rethrow it as an AssertionError because otherwise it pollutes the type of the function with a throws clause even though UTF-8 is a required encoding on all conforming implementations of the JVM.

dwightguth commented on kernel/src/main/java/org/kframework/parser/json/JsonParser.java:
```diff
@@ -0,0 +1,103 @@
+            case KAPPLY:
+                int arity = data.getInt("arity");
+                K[] args  = toKs(arity, data.getJsonArray("args"));
+                return KApply.of(KLabel(data.getString("label")), args);
```
This does not correctly handle the case when the klabel is a variable.

dwightguth commented on kernel/src/main/java/org/kframework/parser/json/JsonParser.java:
```diff
@@ -0,0 +1,103 @@
+
+            case INJECTEDKLABEL:
+                return new InjectedKLabel(KLabel(data.getString("name")));
+
```
If you want to be able to serialize and deserialize arbitrary rules, you need to also include a case for KAs.

dwightguth commented on kernel/src/main/java/org/kframework/parser/json/JsonParser.java:
```diff
@@ -0,0 +1,103 @@
+                return new KRewrite(lhs, rhs);
+
+            case INJECTEDKLABEL:
+                return new InjectedKLabel(KLabel(data.getString("name")));
```
This also does not correctly handle variables.

dwightguth commented on kernel/src/main/java/org/kframework/unparser/ToJson.java:
```diff
@@ -0,0 +1,125 @@
+            InjectedKLabel inj = (InjectedKLabel) k;
+
+            knode.add("node", JsonParser.INJECTEDKLABEL);
+            knode.add("name", inj.klabel().name());
```
This also needs a variable field.

dwightguth commented on kernel/src/main/java/org/kframework/unparser/ToJson.java:
```diff
@@ -0,0 +1,125 @@
+            knode.add("node", JsonParser.INJECTEDKLABEL);
+            knode.add("name", inj.klabel().name());
+
+        }
```
Once again you are missing the case for KAs.

dwightguth reviewed (changes requested):
Looks pretty good but there are some tricky corner cases.

Commit:
kernel/.../{JsonParser,KPrintTest}: throw AssertionError instead of UnsupportedEncodingException

Commit:
kernel/.../{JsonParser,ToJson}: add cases for KAs

ehildenb commented:
@dwightguth I believe I've addressed your concerns:

-   I've fixed the `UnsupportedEncodingException => AssertionError`.
-   I've added both Json serializer/deserializer for `KAs` case.
-   I've handled the cases where a `KLabel` can also be a `KVariable`.

Commit:
kernel/.../{JsonParser,ToJson}: handle cases where KLabel can be a KVariable

Commit:
Merge branch 'master' into json-output

ehildenb commented:
@dwightguth feel free to merge when you feel ready, I still do not have merge permissions.
